metron-platform/metron-parsers/src/main/java/org/apache/metron/parsers/cef/CEFParser.java

```diff
@@ -0,0 +1,274 @@
+		sb.append("\\|");
+		headerBlock("Name", sb);
+		sb.append("\\|");
+		headerBlock("Severity", sb);
+		sb.append("\\|");
+
+		// extension capture:
+		sb.append("(?<extensions>.*)");
+		String pattern = sb.toString();
+
+		p = Pattern.compile(pattern);
+
+		// key finder for extensions
+		pext = Pattern.compile(EXTENSION_CAPTURE_PATTERN);
+	}
+
+	@SuppressWarnings("unchecked")
+	public List<JSONObject> parse(byte[] rawMessage) {
+		List<JSONObject> messages = new ArrayList<>();
+
+		String cefString = new String(rawMessage, UTF_8);
+
+		Matcher matcher = p.matcher(cefString);
+
+		while (matcher.find()) {
+			JSONObject obj = new JSONObject();
+			if (matcher.matches()) {
+				LOG.info(String.format("Found %d groups", matcher.groupCount()));
+				obj.put("DeviceVendor", matcher.group("DeviceVendor"));
+				obj.put("DeviceProduct", matcher.group("DeviceProduct"));
+				obj.put("DeviceVersion", matcher.group("DeviceVersion"));
+				obj.put("DeviceEvent", matcher.group("DeviceEvent"));
+				obj.put("Name", matcher.group("Name"));
+				obj.put("Severity", standardizeSeverity(matcher.group("Severity")));
+			}
+
+			String ext = matcher.group("extensions");
+			Matcher m = pext.matcher(ext);
+
+			int index = 0;
+			String key = null;
+			String value = null;
+			Map<String, String> labelMap = new HashMap<String, String>();
+
+			while (m.find()) {
+				if (key == null) {
+					key = ext.substring(index, m.start());
+					index = m.end();
+					if (!m.find()) {
+						break;
+					}
+				}
+				value = ext.substring(index, m.start());
+				index = m.end();
+				int v = value.lastIndexOf(" ");
+				if (v > 0) {
+					String temp = value.substring(0, v).trim();
+					if (key.endsWith("Label")) {
+						labelMap.put(key.substring(0, key.length() - 5), temp);
+					} else {
+						obj.put(key, temp);
+					}
+					key = value.substring(v).trim();
+				}
+			}
+			value = ext.substring(index);
+
+			// Build a map of Label extensions to apply later
+			if (key.endsWith("Label")) {
+				labelMap.put(key.substring(0, key.length() - 5), value);
+			} else {
+				obj.put(key, value);
+			}
+
+			// Apply the labels to custom fields
+			for (Entry<String, String> label : labelMap.entrySet()) {
+				mutate(obj, label.getKey(), label.getValue());
+			}
+
+			// Rename standard CEF fields to comply with Metron standards
+			obj = mutate(obj, "dst", "ip_dst_addr");
+			obj = mutate(obj, "dpt", "ip_dst_port");
+			obj = convertToInt(obj, "ip_dst_port");
+
+			obj = mutate(obj, "src", "ip_src_addr");
+			obj = mutate(obj, "spt", "ip_src_port");
+			obj = convertToInt(obj, "ip_src_port");
+
+			obj = mutate(obj, "act", "deviceAction");
+			// applicationProtocol
+			obj = mutate(obj, "app", "protocol");
+
+			obj.put("original_string", cefString);
+
+			// apply timestamp from message if present, using rt, syslog
+			// timestamp,
+			// default to current system time
+
+			if (obj.containsKey("rt")) {
+				String rt = (String) obj.get("rt");
```

Comment:
The CEF spec provides for rt (and other datetime fields) to be in epoch milliseconds. Can we add support for that as well?